cumulusci/robotframework/locator_manager.py

```diff
@@ -0,0 +1,149 @@
+
+    | page should contain element  example:dialog.button:Save
+
+    ... will result in "//button[text()='Save']" being passed
+    to 'page should contain element'
+
+"""
+
+LOCATORS = {}
+
+
+def register_locators(prefix, locators):
+    """Register a strategy with a set of locators or a keyword
+
+    If the prefix is already known, merge in the new locators.
+    """
+    if prefix in LOCATORS:
+        BuiltIn().log(f"merging keywords for prefix {prefix}", "DEBUG")
+        dictmerge(LOCATORS[prefix], locators)
+    else:
+        BuiltIn().log(f"registering keywords for prefix {prefix}", "DEBUG")
+        LOCATORS[prefix] = locators
+
+
+def add_location_strategies():
+    """Call selenium's add_location_strategy keyword for all strategies"""
+
+    # selenium throws an error if the location strategy already
+    # exists, so we use a flag to make sure this code is called
+    # only once.
+    selenium = BuiltIn().get_library_instance("SeleniumLibrary")
+    for (prefix, strategy) in LOCATORS.items():
+        try:
+            BuiltIn().log(f"adding location strategy for '{prefix}'", "DEBUG")
+            if isinstance(strategy, dict):
+
+                def strategy(parent, locator, tag, constraints):
+                    locate_element(prefix, parent, locator, tag, constraints)
+
+                selenium.add_location_strategy(
+                    prefix, functools.partial(locate_element, prefix)
+                )
+            else:
+                # not a dict? Just pass it through to selenium as-is
+                # so that this function can register normal keywords
+                selenium.add_location_strategy(prefix, strategy)
+        except Exception as e:
+            BuiltIn().log(f"unable to register locators: {e}", "DEBUG")
+
+
+def locate_element(prefix, parent, locator, tag, constraints):
+    """This is the function called by SeleniumLibrary when a custom locator
+    strategy is used (eg: cci:foo.bar). We pass an additional argument,
+    prefix, so we know which set of locators to use.
+
+    This tokenizes the locator and then does a lookup in the dictionary associated
+    with the given prefix. If any arguments are present, they are applied with
+    .format() before being used to find an element.
+    """
+
+    # Ideally we should call get_webelements (plural) and filter
+    # the results based on the tag and constraints arguments, but
+    # the documentation on those arguments is virtually nil and
+    # SeleniumLibrary's filter mechanism is a private function. In
+    # practice it probably won't matter <shrug>.
+    selenium = BuiltIn().get_library_instance("SeleniumLibrary")
+    loc = translate_locator(prefix, locator)
+    BuiltIn().log(f"locate_element: '{prefix}:{locator}' => {loc}", "DEBUG")
+    try:
+        element = selenium.get_webelement(loc)
+    except Exception:
+        raise Exception(
+            f"Element with locator '{prefix}:{locator}' not found\ntranslated: '{loc}'"
```

Comment:
Including the translated locator here is a nice touch.

Reply:
This helped me find a bug that Selenium was sometimes unregistering the locators when I wasn't expecting it. Unfortunately, if the location strategy is unregistered, robot/selenium will try to use the pattern as a plain text string rather than saying something like "unknown strategy x". This is both helpful when debugging but also acts as a canary in the coal mine for when something goes wrong with the locators.